kafka/kafka.go

```diff
@@ -168,3 +189,73 @@ func (t *topicSync) PublishBatch(ctx context.Context, batch []*bps.PubMessage) e
 	}
 	return t.producer.SendMessages(pmsgs)
 }
+
+// --------------------------------------------------------------------
+
+// Subscriber wraps a kafka consumer and implements the bps.Subscriber interface.
+// It starts consuming from the newest offset (so from message, received after connecting to kafka).
+type Subscriber struct {
+	consumer      sarama.Consumer
+	initialOffset int64
+}
+
+// NewSubscriber inits a new subscriber.
+func NewSubscriber(addrs []string, config *sarama.Config) (*Subscriber, error) {
+	consumer, err := sarama.NewConsumer(addrs, config)
+	if err != nil {
+		return nil, err
+	}
+	return &Subscriber{
+		consumer:      consumer,
+		initialOffset: config.Consumer.Offsets.Initial,
+	}, nil
+}
+
+// Subscribe implements the bps.Subscriber interface.
+func (s *Subscriber) Subscribe(ctx context.Context, topic string, handler bps.Handler) error {
+	parts, err := s.consumer.Partitions(topic)
+	if err != nil {
+		return fmt.Errorf("get %s partitions: %w", topic, err)
+	}
+
+	group, ctx := errgroup.WithContext(ctx)
+	for i := range parts {
+		part := parts[i]
+		group.Go(func() error {
+			csm, err := s.consumer.ConsumePartition(topic, part, s.initialOffset)
+			if err != nil {
+				return fmt.Errorf("consume %s/%d partition: %w", topic, part, err)
+			}
+			defer csm.Close()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+
+				case msg, more := <-csm.Messages():
+					if !more {
+						return nil
+					}
+					// TODO: should we bother with locking/channeling messages for handler?
+					//       Or just demand bps.Handler to be thread safe?
+					if err := handler.Handle(bps.RawSubMessage(msg.Value)); errors.Is(err, bps.Done) {
+						return nil
+					} else if err != nil {
+						return err
+					}
+
+					// TODO: would be nice to log errors as well:
```

Comment:
an `ErrorHandler func(err error) (stop bool)` could be a subscriber option (later)